### PR TITLE
New version of the AnDO

### PR DESCRIPTION
**Pull request by Slowblitz**
Here is the first draft of the Ando that check the new directory named "ephys".

### CODE REVIEW

File: ando/error.py [hunk 1]
```diff
@@ -92,24 +92,23 @@ def __init__(self, arg):
                 + '<li>"BBBB" is a string freely usable by the research group / user , this string cannot contain the underscore character.</li>'\
                 + '<ul></div></div>'
 
+class EphysError(Exception):
+    """Exception raised when the name does not follow the AnDO specification of subject level
```
Comment (JuliaSprenger) [inline ando/error.py:96]:
```suggestion
    """Exception raised when the name does not follow the AnDO specification on the ephy level
```
?

File: ando/error.py [hunk 2]
```diff
@@ -92,24 +92,23 @@ def __init__(self, arg):
...
-                      '1 error found at data folder level.  </div><div class="card-body"> ' \
-                      '<h4 class="em-header clearfix"><strong class="em-header pull-left">' \
-                      f'Error at level 4 [data folder error] at : {names[2]} </strong>' \
-                      f'</h4><br> A subfolder <i><b>{type}</i></b> is required in the session ' \
-                      f'folder "{names[2]}".</div></div>'
+                + '1 error found at session folder level.  </div><div class="card-body"> ' \
+                + '<h4 class="em-header clearfix"><strong class="em-header pull-left">Error 4 type ' \
+                + "[Subject ephys error] at : "+names[2]+" </strong></h4><br><b><i>" \
+                + '</b></i>Ephys folder is missing or does not follow the AnDO specification rules:  '\
+                + ' <ul><li> ephys</li><ul></div></div>'
```
Comment (JuliaSprenger) [inline ando/error.py:113]: Is there a particular reason you are trying to avoid the python 3 string formatting via `f'{names[2]}'`?

File: ando/error.py (hunk 1 above)
Comment (Slowblitz) [reply ando/error.py]: Yup ..

File: ando/tests/datasetDescription.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+| dataset number |      Errors   |
+|:----------:|:-------------:|
+| 001 |  Experiment error |
+| 002 |   Subject error   |
```
Comment (JuliaSprenger) [inline ando/tests/datasetDescription.md:4]:
```suggestion
| 002 |  Subject error  | 
```

File: ando/tests/test_Rules.py [hunk 4]
```diff
@@ -48,6 +48,15 @@ def test_subject_rules_2(self):
         list = ["Sub_001"]
         self.assertEqual(andoE.is_subject(list), False)
 
+    '''testing ephys rules'''
+    def test_ephys_rules(self):
+        list = ["ephys"]
+        self.assertEqual(andoE.is_ephys(list), True)
+    def test_ephys_rules1(self):
+        list = ["Ephys"]
+        self.assertEqual(andoE.is_ephys(list), False)
```
Comment (JuliaSprenger) [inline ando/tests/test_Rules.py:57]: maybe add one test with a string that just contains 'ephys' but also other characters?

File: tools/generator/AnDOGenerator.py [hunk 5]
```diff
@@ -123,14 +123,14 @@ def __init__(self, expName=None, guid=None, sesID=None, date=None, sesNumber=Non
     def get_session_path(self):
         path = os.path.join(f'exp-{self.expName}',
                             f'sub-{self.guid}',
-                            f'ses-{self.sesID}')
+                            f'ses-{self.sesID}',
+                            f'ephys')
```
Comment (JuliaSprenger) [inline tools/generator/AnDOGenerator.py:127]: I am not sure that this would be the right place for this change, as this is class is still intended for folder handling on the session level. Maybe revert this change here and just replace the datafolder list below for now?

File: tools/generator/AnDOGenerator.py [hunk 6]
```diff
@@ -123,14 +123,14 @@ def __init__(self, expName=None, guid=None, sesID=None, date=None, sesNumber=Non
...
         return path
 
     def get_all_folder_paths(self):
         paths = []
         session = self.get_session_path()
-        for datafolder in ['rawdata', 'metadata', 'derivatives']:
-            paths.append(os.path.join(session, datafolder))
+        paths.append(session)
```
Comment (JuliaSprenger) [inline tools/generator/AnDOGenerator.py:133]: keep the datafolder list here for now.

File: tools/generator/tests/test_AnDOGenerator.py [hunk 7]
```diff
@@ -79,17 +80,16 @@ def test_complex_parameters(self):
...
-        expected = os.path.join(f'exp-{self.expName}', f'sub-{self.guid}', f'ses-{self.sesID}')
+        expected = os.path.join(f'exp-{self.expName}', f'sub-{self.guid}', f'ses-{self.sesID}',f'{self.ephys}')
         self.assertEqual(expected, ses.get_session_path())
-
-        self.assertEqual(3, len(ses.get_all_folder_paths()))
+        self.assertEqual(1, len(ses.get_all_folder_paths())) # There is now only one subfolder called ephys
 
     def test_generate_folders(self):
-        ses = AnDOSession(self.expName, self.guid, self.sesID)
+        ses = AnDOSession(self.expName, self.guid, self.sesID,self.ephys)
```
Comment (JuliaSprenger) [inline tools/generator/tests/test_AnDOGenerator.py:92]: This additional argument is not implemented as far as I can see and it also does not really make sense, since the 'ephy' part is fixed for now in the new BEP....

File: tools/generator/tests/test_AnDOGenerator.py [hunk 8]
```diff
@@ -79,17 +80,16 @@ def test_complex_parameters(self):
         self.assertEqual(self.guid, ses.guid)
         self.assertEqual(self.sesID, str(ses.sesID))
 
-    # TODO: This check is still failing, review engine/check_Path for consistency
+
     def test_paths(self):
-        ses = AnDOSession(self.expName, self.guid, self.sesID)
+        ses = AnDOSession(self.expName, self.guid, self.sesID, self.ephys)
```
Comment (JuliaSprenger) [inline tools/generator/tests/test_AnDOGenerator.py:85]: This additional argument is not implemented as far as I can see and it also does not really make sense, since the 'ephy' part is fixed for now in the new BEP....

File: ando/error.py (hunk 2 above)
Comment (Slowblitz) [reply ando/error.py:113]: No, just old style that I'm trying to get rid of =)

File: ando/tests/test_Rules.py (hunk 4 above)
Comment (Slowblitz) [reply ando/tests/test_Rules.py:57]: Yep that's a good Idea.

File: tools/generator/AnDOGenerator.py (hunk 5 above)
Comment (Slowblitz) [reply tools/generator/AnDOGenerator.py]: OK, reverted and replacing the list below.

File: tools/generator/AnDOGenerator.py (hunk 6 above)
Comment (Slowblitz) [reply tools/generator/AnDOGenerator.py]: Updated now datafolder list just contain "ephys"

File: tools/generator/tests/test_AnDOGenerator.py (hunk 7 above)
Comment (Slowblitz) [reply tools/generator/tests/test_AnDOGenerator.py]: Reverted

File: tools/generator/tests/test_AnDOGenerator.py (hunk 8 above)
Comment (Slowblitz) [reply tools/generator/tests/test_AnDOGenerator.py]: Reverted

File: tools/generator/tests/test_AnDOGenerator.py (hunk 7 above)
Comment (Slowblitz) [reply tools/generator/tests/test_AnDOGenerator.py]: but we still need to add it at the end like :
``` ses = AnDOSession(self.expName, self.guid, self.sesID,"ephys")```
Right ?

Comment (JuliaSprenger) [reply tools/generator/tests/test_AnDOGenerator.py]: Yes we can, but I wouldn't since it's not an actual variable parameter, but fixed for us. Maybe it would make sense to define the string 'ephys' as a kind of global variable within the AnDO generator?

Comment (SylvainTakerkart) [reply tools/generator/tests/test_AnDOGenerator.py]: I'm not sure about a global variable, since we will soon want to extend this to other modalities like "vsdi"... so we'd need a solution that could handle both...

although I understand that the checker will not work exactly the same way for ephys and vsdi (since the required metadata might not be the same)... I guess we probably need to have a mechanism that checks for different things according to the modality, no? but we don't have to implement this mechanism right now... just think about it for this specific case of handling the 'ephys' string is probably enough for now!!

Comment (SylvainTakerkart) [reply tools/generator/tests/test_AnDOGenerator.py]: ...or maybe a global variable for "ephys" is the best / most simple solution for now! and we'll see how to extend this to "vsdi" later... ;)

Comment (Slowblitz) [reply tools/generator/tests/test_AnDOGenerator.py]: Ok so what do you think if we add in the AnDO generator : 
```python
MANDATOORY_SUBFOLDER = ["ephys"]
```

then in : 

```python
def get_session_path(self):
        path = os.path.join(f'exp-{self.expName}',
                            f'sub-{self.guid}',
                            f'ses-{self.sesID}',
                            folder for folder in MANDATOORY_SUBFOLDER
                            )
```
This should work right ?

Comment (SylvainTakerkart) [reply tools/generator/tests/test_AnDOGenerator.py]: no need for it to be a list... just use
`MANDATORY_SUBFOLDER = "ephys"`
(with only one O at mandatOOry ;) )

WDYT @JuliaSprenger

Comment (JuliaSprenger) [reply tools/generator/tests/test_AnDOGenerator.py]: Sounds good to me!

File: tools/generator/AnDOGenerator.py [hunk 9]
```diff
@@ -123,14 +123,15 @@ def __init__(self, expName=None, guid=None, sesID=None, date=None, sesNumber=Non
     def get_session_path(self):
         path = os.path.join(f'exp-{self.expName}',
                             f'sub-{self.guid}',
-                            f'ses-{self.sesID}'
+                            f'ses-{self.sesID}',
+                            MANDATORY_SUBFOLDER
```
Comment (JuliaSprenger) [inline tools/generator/AnDOGenerator.py:127]: you only need to add the folder below not here, since `get_all_folder_paths` is calling `get_session_path`. Now you would have two nested `ephy` subfolders.

File: tools/generator/tests/test_AnDOGenerator.py [hunk 10]
```diff
@@ -83,7 +82,7 @@ def test_complex_parameters(self):
 
     def test_paths(self):
         ses = AnDOSession(self.expName, self.guid, self.sesID)
-
+        print(ses)
```
Comment (JuliaSprenger) [inline tools/generator/tests/test_AnDOGenerator.py:85]: I guess this is still a leftover from debugging? Shouldn't stay in the codebase.

File: tools/generator/AnDOGenerator.py [hunk 11]
```diff
@@ -142,7 +143,7 @@ def get_all_folder_paths(self):
         assert not check_Path(combined_paths, verbose=False)[0], \
             'Error in AnDO path generation. Generated paths are not consistent with AnDO ' \
             'specifications'
-
+        print(paths)
```
Comment (JuliaSprenger) [inline tools/generator/AnDOGenerator.py:146]: same here: I guess this is still a leftover from debugging? Shouldn't stay in the codebase.

File: tools/generator/AnDOGenerator.py (hunk 9 above)
Comment (Slowblitz) [reply tools/generator/AnDOGenerator.py:127]: It doesn't seem to pass the test_paths function if I remove it :
image
So maybe there is something that I don't understand ?

Comment (JuliaSprenger) [reply tools/generator/AnDOGenerator.py:127]: Yes, but this is because you introduced 'ephys' also in the expected result. Before this PR it was just generating the session_path (so everything down to the session level, without the data folders).

Comment (Slowblitz) [reply tools/generator/AnDOGenerator.py:127]: So should I delete the "ephys" in the expected result ?

File: tools/generator/tests/test_AnDOGenerator.py (hunk 10 above)
Comment (Slowblitz) [reply tools/generator/tests/test_AnDOGenerator.py]: deleted

File: tools/generator/AnDOGenerator.py (hunk 11 above)
Comment (Slowblitz) [reply tools/generator/AnDOGenerator.py]: deleted

**Comment by JuliaSprenger** [comment]
Ok, so this now simplified the folder structure. II think that is all we planned for this PR. So I think we can merge soon.

**Review by SylvainTakerkart — CHANGES_REQUESTED** [review]
Hello,

I trust @JuliaSprenger on the review of the code itself... On my side, I tried to execute the CLI...

`
> cd AnDO
>git checkout enh
> cd ando
> python checker.py tests/dataset007/exp-Landing/

`

and I get an error:

`
> Traceback (most recent call last):
>   File "checker.py", line 87, in <module>
>     main()
>   File "checker.py", line 75, in main
>     error_found = is_AnDO(directory, False, False)
> TypeError: is_AnDO() takes 1 positional argument but 3 were given
`

Is this normal?

File: tools/generator/tests/test_AnDOGenerator.py [hunk 12]
```diff
@@ -52,6 +51,7 @@ def setUp(self):
         self.sesNumber = '100'
         self.customSesField = 'test'
         self.sesID = f'{self.date}_{self.sesNumber}_{self.customSesField}'
+        self.ephys = 'ephys'
```
Comment (JuliaSprenger) [inline tools/generator/tests/test_AnDOGenerator.py:54]: I don't think you are using this anywhere. Better remove it then.

Comment (Slowblitz) [reply tools/generator/tests/test_AnDOGenerator.py]: Done =)

**Comment by SylvainTakerkart** [comment]
> Hello,
>
> 
> Is this normal?

answering my own question... that error came from having an old version of AnDO still installed on my system (and a non-trivial call of that old version)... so now, everything is fine for me!

**Comment by JuliaSprenger** [comment]
Ok, with all comments being resolved, I am going to merge this one now.
